### PR TITLE
Feature - Recover Wallet CLI

### PR DESCRIPTION
**Pull request by dannypaz**
## Description
With this PR, the user of a broker, with a new instance of the broker, can restore on-chain AND committed funds from the broker's wallet.

`sparkswap wallet recover <symbol>`

The user will be prompted for their original wallet password, 24 word recovery seed and will be asked if they want to use a backup file (backup file will need to be located at `./backup/lnd_<symbol>.backup`)

## Related PRs
List related PRs if applicable

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Link to Trello

### CODE REVIEW

File: broker-cli/commands/wallet.js [hunk 1]
```diff
@@ -529,6 +530,35 @@ async function unlock (args, opts, logger) {
...
+ */
+async function recover (args, opts, logger) {
+  const { symbol } = args
+  const { rpcAddress = null } = opts
+
+  try {
+    const client = new BrokerDaemonClient(rpcAddress)
+
+    const password = await askQuestion(`Please enter your wallet password:`, { silent: true })
+    const seed = await askQuestion(`Please enter your 24 word recover seed, with each word separated by spaces`, { silent: true })
```
Comment (treygriffith) [inline broker-cli/commands/wallet.js]: `silent: true` seems pretty rough for this if it does what I think it does.

File: broker-cli/commands/wallet.js [hunk 2]
```diff
@@ -529,6 +530,35 @@ async function unlock (args, opts, logger) {
...
+    const backupPath = await askQuestion(`Enter the path to your backup file (optional)`)
```
Comment (treygriffith) [inline broker-cli/commands/wallet.js]: Why is this optional?

File: broker-cli/proto/broker.proto [hunk 3]
```diff
@@ -1787,6 +1787,21 @@ message ChangeWalletPasswordRequest {
...
+ * NOTE: This can only be done on a fresh install of the broker
+ */
+message RecoverWalletRequest {
+  // Currency of wallet
+  string symbol = 1;
+  // Password from the existing wallet
+  string password = 2;
+  // 24 word mnemonic seed
+  repeated string seed = 3;
+  // The path of the backup file (path inside of the broker container)
```
Comment (treygriffith) [inline broker-cli/proto/broker.proto]: a path inside the broker container is pretty confusing.

File: broker-daemon/broker-rpc/wallet-service/recover-wallet.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  } = params
+
+  const engine = engines.get(symbol)
+
+  if (!engine) {
+    logger.error(`Could not find engine: ${symbol}`)
+    throw new Error(`Unable to create wallet for engine: ${symbol}`)
+  }
+
+  console.log(params)
```
Comment (treygriffith) [inline broker-daemon/broker-rpc/wallet-service/recover-wallet.js]: i'm assuming this is still wip left over

File: broker-daemon/utils/migrate-store.spec.js [hunk 5]
```diff
@@ -1,117 +1,201 @@
-const { sinon, expect } = require('test/test-helper')
+const path = require('path')
+const {
+  sinon,
+  expect,
+  rewire
+} = require('test/test-helper')
 
-const migrateStore = require('./migrate-store')
+const migrateStore = rewire(path.resolve(__dirname, './migrate-store'))
```
Comment (treygriffith) [inline broker-daemon/utils/migrate-store.spec.js]: did you base this off another branch? Including `migrateStore` in this PR seems quite strange.

Comment (dannypaz) [reply broker-daemon/utils/migrate-store.spec.js]: @treygriffith im not sure why the dif was weird

File: broker-cli/commands/wallet.js (hunk 2 above)
Comment (dannypaz) [reply broker-cli/commands/wallet.js]: from the CLI, we should force the user to provide this. I haven't made the change yet

File: broker-cli/commands/wallet.js [hunk 6]
```diff
@@ -529,6 +530,48 @@ async function unlock (args, opts, logger) {
   }
 }
 
+/**
+ * Recover a wallet from a provided mnemonic seed and password. backup file is optional
```
Comment (treygriffith) [inline broker-cli/commands/wallet.js]: [nit] More information about the backup file (i.e. off-chain funds) would be nice.

File: broker-cli/commands/wallet.js [hunk 7]
```diff
@@ -529,6 +530,48 @@ async function unlock (args, opts, logger) {
...
+async function recover (args, opts, logger) {
+  const { symbol } = args
+  const { rpcAddress = null } = opts
+
+  try {
+    const client = new BrokerDaemonClient(rpcAddress)
+
+    const password = await askQuestion(`Please enter your wallet password:`, { silent: true })
+    let seed = await askQuestion('Please enter your 24 word recovery seed, with each word separated by spaces:')
+    const shouldUseBackup = await askQuestion('Recover with backup file? Y/N:')
```
Comment (treygriffith) [inline broker-cli/commands/wallet.js]: `Recover committed funds with backup file?` would be better I think?

File: broker-cli/proto/broker.proto [hunk 8]
```diff
@@ -1787,6 +1787,20 @@ message ChangeWalletPasswordRequest {
...
+ * Request to recover a wallet
+ */
+message RecoverWalletRequest {
+  // Currency of wallet
+  string symbol = 1;
+  // Password from the existing wallet
+  string password = 2;
+  // 24 word mnemonic seed
+  repeated string seed = 3;
+  // The path of the backup file
```
Comment (treygriffith) [inline broker-cli/proto/broker.proto]: This doc is out of date.

File: broker-cli/proto/broker.proto [hunk 9]
```diff
@@ -1787,6 +1787,20 @@ message ChangeWalletPasswordRequest {
...
+  bool backup = 4;
```
Comment (treygriffith) [inline broker-cli/proto/broker.proto]: `use_backup` might be better. Also, maybe something to make it more generic, like `recover_committed`?

Comment (dannypaz) [reply broker-cli/proto/broker.proto]: I agree the name can change. I like `recover-committed`, but going to think on it for a bit

File: broker-cli/proto/broker.proto [hunk 10]
```diff
@@ -1787,6 +1787,20 @@ message ChangeWalletPasswordRequest {
...
+ * Request to recover a wallet
+ */
+message RecoverWalletRequest {
+  // Currency of wallet
+  string symbol = 1;
+  // Password from the existing wallet
+  string password = 2;
+  // 24 word mnemonic seed
+  repeated string seed = 3;
+  // True/False signifying the use of a backup file for committed funds
```
Comment (treygriffith) [inline broker-cli/proto/broker.proto]: is this too tied to LND?

Comment (dannypaz) [reply broker-cli/proto/broker.proto]: That is a good Q. I made a few changes to mitigate this. We'll refer to this flag as `use_backup` not backup file and i've updated documentation to specify when something is related to lnd or not.

File: broker-cli/proto/broker.proto [hunk 11]
```diff
@@ -2168,4 +2182,58 @@ service WalletService {
...
+   * 3. A backup file (optional, required for off-chain recovery)
+   *
+   * *** On-chain recovery
+   *
+   * This RPC allows the user to recover any funds that have NOT been committed to the exchange with the
+   * use of only your wallet password and 24-word recovery seed.
+   *
+   * *** Off-chain recovery
+   *
+   * When a backup is present, this RPC allows the user to recover any funds that were deposited to the exchange, as well as
```
Comment (treygriffith) [inline broker-cli/proto/broker.proto]: `deposited` is not the right word here. What are you trying to say?

File: broker-cli/utils/ask-question.js [hunk 12]
```diff
@@ -59,16 +59,28 @@ function askQuestion (message, { silent = false } = {}) {
...
-        rl.history = rl.history.slice(1)
+        // Remove the data listener that we had added so that we are not adding
+        // multiple listeners on stdin which causes weird side-effects when using
+        // `askQuestion` multiple times in the same command
+        process.stdin.removeListener('data', dataHandler)
+
+        // If the process is interactive (non-TTY) then we can remove the history.
+        // If the terminal is TTY then rl.history is not defined
+        if (rl.history) {
+          rl.history = rl.history.slice(1)
```
Comment (treygriffith) [inline broker-cli/utils/ask-question.js]: can we do `historySize` instead?

**Pull request merged**